### PR TITLE
Dataloader

### PR DESCRIPTION
**Pull request by zaitmambet**

### CODE REVIEW

**Comment by KernelA** [comment]
@zaitmambet , зачем в Jupyter notebook как я это буду использовать в коде обучения модели?

**Review by KernelA — CHANGES_REQUESTED** [review]
Нужно переделать, чтобы я мог использовать код в основном проекте. Я не смогу из Jupyter импортировать методы и классы в Python

File: dataloader/data_loader.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    def __getitem__(self, index):
+        with zipfile.ZipFile(self._path_to_zip) as zip_file:
+            with zip_file.open(self._image_lists[index], "r") as image_file:
+                image = np.frombuffer(image_file.read(), dtype=np.uint8)
+                image = io.decode_jpeg(torch.from_numpy(image.copy()), mode=io.ImageReadMode.RGB)
+
+        return {"image": self._transform(image), "label": self._image_classes[index]}
+
+
+class XrayDataLoader(utils.data.DataLoader):
```
Comment (KernelA) [inline dataloader/data_loader.py:104]: Зачем свой Dataloader?

Стандартный и так пойдёт

**Comment by KernelA** [comment]
Зачем тут два раза добавлены notebook c одним названием?

File: dataloader/data_loader.py (hunk 1 above)
Comment (zaitmambet) [reply dataloader/data_loader.py:104]: наверное ты прав, единственное надо как-то передать место хранения данных и их тип (трейн, тест, вал)

File: dataloader/data_loader.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    h_padding = (max_wh - w) / 2
+    v_padding = (max_wh - h) / 2
+    l_pad = h_padding if h_padding % 1 == 0 else h_padding+0.5
+    t_pad = v_padding if v_padding % 1 == 0 else v_padding+0.5
+    r_pad = h_padding if h_padding % 1 == 0 else h_padding-0.5
+    b_pad = v_padding if v_padding % 1 == 0 else v_padding-0.5
+    padding = (int(l_pad), int(t_pad), int(r_pad), int(b_pad))
+    return padding
+
+class NewPad(object):
```
Comment (KernelA) [inline dataloader/data_loader.py:33]: Зачем тут своё преобразование, когда есть это?
https://pytorch.org/vision/stable/transforms.html#torchvision.transforms.Pad

File: dataloader/data_loader.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            PIL Image: Padded image.
+        """
+        return F.pad(img, get_padding(img), self.fill, self.padding_mode)
+    
+    def __repr__(self):
+        return self.__class__.__name__ + '(padding={0}, fill={1}, padding_mode={2})'.\
+            format(self.fill, self.padding_mode)
+
+
+data_transforms = {
```
Comment (KernelA) [inline dataloader/data_loader.py:56]: Это нужно вынести в отдельный файл и передавать как параметр в Dataloader.

Должно быть хотя бы два типа для train и test

File: dataloader/data_loader.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    'train': transforms.Compose([
+        transforms.RandomHorizontalFlip()
+    ]),
+    'general': transforms.Compose([
+        #NewPad(),
+        transforms.Resize(size=500),
+        transforms.ConvertImageDtype(dtype=torch.float32), 
+        transforms.Grayscale(num_output_channels=3),
```
Comment (KernelA) [inline dataloader/data_loader.py:64]: Зачем это нужно оно и так в Grayscale?

File: dataloader/data_loader.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+class ChestXray(utils.data.Dataset):
+    def __init__(self, path_to_zip, set_name, class_names = {"NORMAL": 0, "PNEUMONIA": 1}):
+        self._path_to_zip = path_to_zip
+        self._class_mapping = class_names
+        self._image_lists = []
+        self._image_classes = []
+        self._set_name = set_name
+        assert set_name in ['train', 'test', 'val']
+        if set_name == 'train':
+          self._transform = transforms.Compose([data_transforms['general'], data_transforms['train']])
```
Comment (KernelA) [inline dataloader/data_loader.py:79]: Это надо в параметрах принимать

File: dataloader/data_loader.py (hunk 2 above)
Comment (zaitmambet) [reply dataloader/data_loader.py:33]: здесь используется функция get_padding. Но в общем, NewPad() я далее в data_transforms не стал использовать, так что можно убрать

**Review by KernelA — CHANGES_REQUESTED** [review]
Нужно доделать

File: dataloader/data_loader.py (hunk 4 above)
Comment (zaitmambet) [reply dataloader/data_loader.py:64]: ты разве не переводил в RGB? проверю

File: dataloader/data_loader.py (hunk 1 above)
Comment (KernelA) [reply dataloader/data_loader.py:104]: Тут стандартный всё делает. Нам нет смысла делать свой

File: dataloader/data_loader.py (hunk 5 above)
Comment (zaitmambet) [reply dataloader/data_loader.py:79]: кстати да, не заметил!

File: dataloader/data_loader.py (hunk 2 above)
Comment (KernelA) [reply dataloader/data_loader.py:33]: Надо убрать всё лишнее

File: dataloader/data_loader.py (hunk 5 above)
Comment (zaitmambet) [reply dataloader/data_loader.py:79]: хотя там же выше от параметра зависит: if set_name == 'train':

File: dataloader/data_loader.py (hunk 4 above)
Comment (KernelA) [reply dataloader/data_loader.py:64]: В том то и дело что при загрузке уже RGB

File: dataloader/data_loader.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            format(self.fill, self.padding_mode)
+
+
+data_transforms = {
+    'train': transforms.Compose([
+        transforms.RandomHorizontalFlip()
+    ]),
+    'general': transforms.Compose([
+        #NewPad(),
+        transforms.Resize(size=500),
```
Comment (KernelA) [inline dataloader/data_loader.py:62]: Это должно быть как параметр

File: dataloader/data_loader.py (hunk 5 above)
Comment (KernelA) [reply dataloader/data_loader.py:79]: Я про сам transform он должен извне передаваться. Нам потом на этапе применения модели нужно иметь возможность использовать отдельно. Если он будет внутри кода глубоко, то мы не сможем использовать

**Comment by KernelA** [comment]
[Сделать EDA для изображений](https://trello.com/c/XwnquiJD/17-%D1%81%D0%B4%D0%B5%D0%BB%D0%B0%D1%82%D1%8C-eda-%D0%B4%D0%BB%D1%8F-%D0%B8%D0%B7%D0%BE%D0%B1%D1%80%D0%B0%D0%B6%D0%B5%D0%BD%D0%B8%D0%B9)

**Comment by KernelA** [comment]
[3 Применение следующих групп методов: gradients, signals, attribution methods (9 баллов)](https://trello.com/c/NzkS7JZl/26-%D0%BF%D1%80%D0%B8%D0%BC%D0%B5%D0%BD%D0%B5%D0%BD%D0%B8%D0%B5-%D1%81%D0%BB%D0%B5%D0%B4%D1%83%D1%8E%D1%89%D0%B8%D1%85-%D0%B3%D1%80%D1%83%D0%BF%D0%BF-%D0%BC%D0%B5%D1%82%D0%BE%D0%B4%D0%BE%D0%B2-gradients-signals-attribution-methods-9-%D0%B1%D0%B0%D0%BB%D0%BB%D0%BE%D0%B2)

**Comment by KernelA** [comment]
[3 Применение следующих групп методов: gradients, signals, attribution methods (9 баллов)](https://trello.com/c/NzkS7JZl/26-%D0%BF%D1%80%D0%B8%D0%BC%D0%B5%D0%BD%D0%B5%D0%BD%D0%B8%D0%B5-%D1%81%D0%BB%D0%B5%D0%B4%D1%83%D1%8E%D1%89%D0%B8%D1%85-%D0%B3%D1%80%D1%83%D0%BF%D0%BF-%D0%BC%D0%B5%D1%82%D0%BE%D0%B4%D0%BE%D0%B2-gradients-signals-attribution-methods-9-%D0%B1%D0%B0%D0%BB%D0%BB%D0%BE%D0%B2)

**Comment by KernelA** [comment]
Все изменения уже есть в другом pull request
